### PR TITLE
Add flattening to MLFLow reporter

### PR DESCRIPTION
**Pull request by ramasesh**
This solves the issue that, previously, MLFlow reporter needed a flat dictionary when reporting params, i.e. a dictionary whose values were convertable to string.  

A common pattern is to save a configuration in nested dictionaries like {"model": {}, "optimization": {}}, etc.  This PR makes it easier to save those, by flattening the dictionary first within the MLFLow reporter.

### CODE REVIEW

**Comment by ramasesh** [comment]
Sam's comments:
Vinay Ramasesh I think you want to do the flattening with this: https://github.com/google/uv-metrics/blob/sritchie%2Fpython3_makefile/uv/util/attachment.py#L68

`by_prefix` should collapse anything nested. one nice change would be to modify it to ignore non-dictionary values

Vinay Ramasesh, 35 mins
But it’s not necessarily a metric

Sam Ritchie, 35 mins
so, make it only collapse the ones we want (and then act recursively on those, so you can nest arbitrarily)

Vinay Ramasesh, 35 mins
The value could just be a strong

String

Sam Ritchie, 35 mins
`Metric = Any`

yeah, a uv Metric is a string, or number, or anything you can serialize

Vinay Ramasesh, 35 mins
Oh ok

Sam Ritchie, 35 mins
it's just a marker type

Vinay Ramasesh, 34 mins
So should I update the PR?

Sam Ritchie, 34 mins
yes please, and add tests

one test of the example you had in the readme is `by_prefix{"model": {}, "optimization": {}} => {}

**Comment by ajslone** [comment]
@ramasesh you're probably doing this already, but running `make pytest` at the top-level directory gives you code coverage info, which really helps when putting in tests

**Comment by ramasesh** [comment]
@sritchie I followed your suggestion to use the `attachment.py` code here, but I figured it was better to make a new functio called `flatten` to do exactly what we want (i.e. ignore key value pairs which are already flat).

My hypothesis testing works but is pretty ugly, so I would welcome feedback there.

File: tests/uv/util/test_attachment.py [hunk 1]
```diff
@@ -39,3 +40,42 @@ def test_by_suffix(m):
...
+
+  assert flattened == m
+
+
+@given(
+    st.dictionaries(st.text(min_size=1),
+                    st.dictionaries(st.text(min_size=1), st.integers())))
+def test_flatten_notflat(m):
+  # Tests that flattening properly handles dictionaries which are not flat
+  flattened = a.flatten(m)
```
Comment (sritchie) [inline tests/uv/util/test_attachment.py:67]: nice, you can also assert `a.flatten(m) == a.flatten(a.flatten(m))`

File: uv/mlflow/reporter.py [hunk 2]
```diff
@@ -18,9 +18,10 @@
 import mlflow as mlf
 from mlflow.entities import Param, Metric, RunTag
 import time
-from typing import Optional, Dict, List
+from typing import Optional, Dict, List, Union
 import uv.reporter.base as b
 import uv.types as t
+from uv.util import attachment
```
Comment (sritchie) [inline uv/mlflow/reporter.py]: I prefer not to do this for modules. Can you do `import uv.util.attachment as ua`, then use `ua.flatten` below?

**Review by sritchie — APPROVED** [review]
Looks fantastic!

**Pull request merged**